### PR TITLE
Add new state in state machine 

### PR DESCRIPTION
**Pull request by gabriel128**
### Problem trying to solve

With the current state machine is not possible to differentiate within a job that has crashed and a job that's taking a long time to run.

### Proposed Solution

Add a RETRYING state  when `queue.perform` fails in the worker, and change re-schedule the job so it can be picked up by the `activate_scheduled_jobs` function. 

For falling back, the worker catches 2 different kind of failures from `queue.perform`, exceptions and transaction failures. 

Thus, with this extra state, it should be possible to pick a RETRYING job and force a retry by changing it to AVAILABLE manually, ensuring that it's not a job that it's taking a long time to run

### New State Machine proposal

ectojob retry state - improved 1

### CODE REVIEW

File: lib/ecto_job/job_queue.ex [hunk 1]
```diff
@@ -307,10 +308,43 @@ defmodule EctoJob.JobQueue do
...
+          where: j.state == "IN_PROGRESS"
+        ),
+        [
+          set: [
+            state: "RETRYING",
+            schedule: progress_time(now, job.attempt + 1, timeout_ms),
+            updated_at: now
+          ]
+        ],
+        returning: true
```
Comment (seangeo) [inline lib/ecto_job/job_queue.ex]: `returning: true` is going to be deprecated in Ecto 3.0.

File: lib/ecto_job/worker.ex [hunk 2]
```diff
@@ -19,14 +19,34 @@ defmodule EctoJob.Worker do
...
+
+  @spec do_work(Config.t, EctoJob.JobQueue.job(), DateTime.t()) ::
+          :ok | {:ok, EctoJob.JobQueue.job()} | {:error, any()}
+  def do_work(config = %Config{repo: repo,
+                               execution_timeout: exec_timeout,
+                               retrying_timeout: retrying_timeout},
+              job,
+              now) do
+    with {:ok, in_progress_job} <- JobQueue.update_job_in_progress(repo, job, now, exec_timeout),
+         {:ok, _} <- run_queue(config, in_progress_job) do
```
Comment (seangeo) [inline lib/ecto_job/worker.ex]: I’m not sure we should force jobs to return this, that changes the contract for a job quite a bit, might be a stretch to get into the main repo. Maybe we should just be catching exceptions to trigger the retrying stage?

File: lib/ecto_job/job_queue.ex (hunk 1 above)
Comment (gabriel128) [reply lib/ecto_job/job_queue.ex]: :+1: I'll change it for `select: j` then

File: lib/ecto_job/worker.ex (hunk 2 above)
Comment (gabriel128) [reply lib/ecto_job/worker.ex]: as I understood, even if it's not explicit in the documentation, and maybe I over assumed it, the last call in the job must be `Repo.transaction()` which returns `{:ok, multi}`. I get what you mean though, and I think you are right.  

Maybe we can loosen up this requirement a bit and do the following:

Comment (gabriel128) [reply lib/ecto_job/worker.ex]:
```
 with {:ok, in_progress_job} <- JobQueue.update_job_in_progress(repo, job, now, exec_timeout),
         response <- run_queue(config, in_progress_job), 
        true <- valid?(resonse) do
      log_duration(config, in_progress_job, now)
      notify_completed(repo, in_progress_job)
....

def valid?({:error, a, b, c}) do  # check that is a multi transaction failure in this one
   false
end

def valid?(_), do: true
```

Comment (gabriel128) [reply lib/ecto_job/worker.ex]: I think that is pretty important to put it in "RETRYING" if it is a transaction failure, even more important than an exception, wdyt?

Comment (gabriel128) [reply lib/ecto_job/worker.ex]: and sorry for splitting the message in 3, pressed `ctrl enter` by mistake hahahah

Comment (seangeo) [reply lib/ecto_job/worker.ex]: Ok cool. Let’s see what they say upstream.

**Review by seangeo — APPROVED** [review]
Let's give it go.

**Comment by gabriel128** [comment]
\o/

File: lib/ecto_job/worker.ex [hunk 3]
```diff
@@ -19,16 +19,44 @@ defmodule EctoJob.Worker do
...
+      error -> error
+    end
+  end
+
+  @spec run_queue(Config.t, EctoJob.JobQueue.job()) :: {:ok, EctoJob.JobQueue.job()} | {:error, any()}
+  defp run_queue(%Config{repo: repo, retrying_timeout: timeout}, job = %queue{}) do
+    try do
+      queue.perform(JobQueue.initial_multi(job), job.params)
+    rescue
+      _ in RuntimeError -> JobQueue.update_job_to_retrying(repo, job, DateTime.utc_now(), timeout)
```
Comment (seangeo) [inline lib/ecto_job/worker.ex:49]: Just had a thought around this, this is going to swallow errors so they’ll never bubble up to rollbar (for us). I think it needs to reraise the error here.

**Pull request merged**